### PR TITLE
Add content, toplevel audit annotations

### LINKED ISSUE #1768: Unable to determine from audit log why workspace access was not permitted
Given an audit event such as (some fields omitted/redacted):
```json
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "Metadata",
  "auditID": "1d1118b4-8190-4629-ae73-a36a11390098",
  "stage": "ResponseComplete",
  "requestURI": "/apis?timeout=32s",
  "verb": "get",
  "user": {
    "username": "[redacted]",
    "groups": [
      "[redacted]",
      "system:authenticated"
    ]
  },
  "sourceIPs": [
    "[redacted]",
    "[redacted]"
  ],
  "userAgent": "kubectl/v1.24.3 (darwin/amd64) kubernetes/aef86a9",
  "responseStatus": {
    "metadata": {},
    "status": "Failure",
    "message": "forbidden: User \"[redacted]\" cannot get path \"/apis\": workspace access not permitted",
    "reason": "Forbidden",
    "details": {},
    "code": 403
  },
  "requestReceivedTimestamp": "2022-08-16T16:11:07.676945Z",
  "stageTimestamp": "2022-08-16T16:11:07.678798Z",
  "annotations": {
    "authorization.k8s.io/decision": "forbid",
    "authorization.k8s.io/reason": "workspace access not permitted",
    "tenancy.kcp.dev/workspace": "root:users:qs:ie:[redacted]:some-child"
  }
}
```

it's not possible to determine why access was not permitted. We need to add more audit annotations to provide additional details.

/assign

### PR DESCRIPTION
**Pull request by ncdc**
## Summary
Add more audit annotations to add details to the audit log when
workspace access is not permitted.

## Related issue(s)

Fixes #1768
Part of #1307

### CODE REVIEW

File: pkg/authorization/toplevel_org_authorizer.go [hunk 1]
```diff
@@ -71,11 +71,13 @@ type topLevelOrgAccessAuthorizer struct {
 func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil || cluster == nil || cluster.Name.Empty() {
+		kaudit.AddAuditAnnotation(ctx, AuditAnnotationWorkspaceAccessNotPermittedDetails, fmt.Sprintf("error getting cluster from request: %v", err))
```
Comment (sttts) [inline pkg/authorization/toplevel_org_authorizer.go:74]: these need another key for this authorizer

File: pkg/authorization/toplevel_org_authorizer.go [hunk 2]
```diff
@@ -105,15 +108,17 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
+			kaudit.AddAuditAnnotation(ctx, AuditAnnotationWorkspaceAccessNotPermittedDetails, fmt.Sprintf("clusterworkspace %q not found", topLevelWSKey))
 			return authorizer.DecisionDeny, WorkspaceAcccessNotPermittedReason, nil
 		}
+		kaudit.AddAuditAnnotation(ctx, AuditAnnotationWorkspaceAccessNotPermittedDetails, fmt.Sprintf("error getting clusterworkspace %q: %v", topLevelWSKey, err))
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, err
 	}
 
-	switch {
-	case isServiceAccount:
+	if isServiceAccount {
```
Comment (sttts) [inline pkg/authorization/toplevel_org_authorizer.go:121]: don't like this rewrite. We intentionally wrote it with a switch to be it clearer. It's not obvious that `isUser != isServiceAccount` but rather a coincidence of the current setup.

Comment (s-urbaniak) [reply pkg/authorization/toplevel_org_authorizer.go:121]: yes agreed, please leave the switch.

Comment (ncdc) [reply pkg/authorization/toplevel_org_authorizer.go:121]: Ok..... but it does make returning harder

**Comment by ncdc** [comment]
@s-urbaniak will do

File: pkg/authorization/toplevel_org_authorizer.go [hunk 3]
```diff
@@ -95,20 +111,42 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
 		if isAuthenticated && (isUser || isServiceAccountFromRootCluster) {
 			return a.delegate.Authorize(ctx, attr)
 		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			TopLevelContentAuditDecision, DecisionNoOpinion,
+			TopLevelContentAuditReason, "only root serviceaccounts may access root",
```
Comment (sttts) [inline pkg/authorization/toplevel_org_authorizer.go:117]: or authenticated user

**Comment by ncdc** [comment]
@s-urbaniak I've made the requested changes for the content and toplevel authorizer audit annotations. This should be enough to debug what I need for now, and we can add more annotations in separate PRs?

File: pkg/authorization/toplevel_org_authorizer.go [hunk 4]
```diff
@@ -71,11 +77,21 @@ type topLevelOrgAccessAuthorizer struct {
...
+		)
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, err
 	}
 
 	if !cluster.Name.HasPrefix(tenancyv1alpha1.RootCluster) {
 		// nobody other than system:masters (excluded from authz) has access to workspaces not based in root
+		kaudit.AddAuditAnnotations(
+			ctx,
+			TopLevelContentAuditDecision, DecisionNoOpinion,
+			TopLevelContentAuditReason, "only system:masters may access workspaces with a non-root prefix",
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:93]: this text is a resolution not a reason. The reason should rather be `toplevel.authorization.kcp.io/decision: non-root prefixed workspace access not permitted`.

File: pkg/authorization/toplevel_org_authorizer.go [hunk 5]
```diff
@@ -95,20 +111,42 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
 		if isAuthenticated && (isUser || isServiceAccountFromRootCluster) {
 			return a.delegate.Authorize(ctx, attr)
 		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			TopLevelContentAuditDecision, DecisionNoOpinion,
+			TopLevelContentAuditReason, "only authenticated users and serviceaccounts in root may access root",
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:117]: same as above, this is a possible resolution, the reason is rather something like `toplevel.authorization.kcp.io/reason: root workspace access by non-root service account not permitted`.

File: pkg/authorization/toplevel_org_authorizer.go [hunk 6]
```diff
@@ -95,20 +111,42 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, nil
 	}
 
 	// get org in the root
 	requestTopLevelOrgName, ok := topLevelOrg(cluster.Name)
 	if !ok {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			TopLevelContentAuditDecision, DecisionNoOpinion,
+			TopLevelContentAuditReason, "not part of root workspace hierarchy",
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:128]: let's add the offending cluster name for debugability: `TopLevelContentAuditReason, fmt.Sprintf("access not permitted, workspace %q is not part of root workspace hierarchy", cluster.Name)`

File: pkg/authorization/toplevel_org_authorizer.go [hunk 7]
```diff
@@ -121,44 +159,56 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
-			}
-			if dec == authorizer.DecisionAllow {
-				return a.delegate.Authorize(ctx, attr)
-			}
+		dec, reason, err := a.rootAuthorizer.Authorize(ctx, workspaceAttr)
+		if err != nil {
+			kaudit.AddAuditAnnotations(
+				ctx,
+				TopLevelContentAuditDecision, DecisionNoOpinion,
+				TopLevelContentAuditReason, fmt.Sprintf("%s (%v)", reason, err))
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:195]: as per https://github.com/kcp-dev/kcp/issues/1307, let's use `TopLevelContentAuditReason, fmt.Sprintf("forbidden by \"content\" verb RBAC in top level workspace %q, reason=%q, error=%v", requestTopLevelOrgName, reason, err))`

File: pkg/authorization/toplevel_org_authorizer.go [hunk 8]
```diff
@@ -121,44 +159,56 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
+			return authorizer.DecisionNoOpinion, reason, err
 		}
-		if len(errList) > 0 {
-			return authorizer.DecisionNoOpinion, strings.Join(reasonList, "\n"), utilerrors.NewAggregate(errList)
+		if dec == authorizer.DecisionAllow {
+			kaudit.AddAuditAnnotations(
+				ctx,
+				TopLevelContentAuditDecision, DecisionAllowed,
+				TopLevelContentAuditReason, fmt.Sprintf("allowed by root authorizer (%s)", reason),
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:202]: similarly:`TopLevelContentAuditReason, fmt.Sprintf("allowed by root workspace RBAC, verb=\"access\" resource=\"workspaces/content\", name=%q, reason=%q", requestTopLevelOrgName, reason))`

File: pkg/authorization/toplevel_org_authorizer.go [hunk 9]
```diff
@@ -121,44 +159,56 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
+			)
+			return a.delegate.Authorize(ctx, attr)
 		}
+
+		kaudit.AddAuditAnnotations(
+			ctx,
+			TopLevelContentAuditDecision, decisionString(dec),
+			TopLevelContentAuditReason, fmt.Sprintf("root authorizer did not allow; reason(s)=%q", reason),
```
Comment (s-urbaniak) [inline pkg/authorization/toplevel_org_authorizer.go:210]: similarly: `TopLevelContentAuditReason, fmt.Sprintf("forbidden by root workspace RBAC, verb=\"access\" resource=\"workspaces/content\", name=%q, reason=%q", requestTopLevelOrgName, reason))`

File: pkg/authorization/workspace_content_authorizer.go [hunk 10]
```diff
@@ -75,10 +87,20 @@ type workspaceContentAuthorizer struct {
...
+			WorkspaceContentAuditReason, fmt.Sprintf("error getting cluster from request: %v", err),
+		)
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, err
 	}
 	// empty or non-root based workspaces have no meaning in the context of authorizing workspace content.
 	if cluster == nil || cluster.Name.Empty() || !cluster.Name.HasPrefix(tenancyv1alpha1.RootCluster) {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			WorkspaceContentAuditDecision, DecisionNoOpinion,
+			WorkspaceContentAuditReason, "empty or non root workspace",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:102]: let's print out the cluster name using %q

File: pkg/authorization/workspace_content_authorizer.go [hunk 11]
```diff
@@ -98,14 +120,31 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
+				WorkspaceContentAuditDecision, DecisionAllowed,
+				WorkspaceContentAuditReason, "subject is either an authenticated user or a serviceaccount in root",
+			)
+
 			return a.delegate.Authorize(ctx, withGroups)
 		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			WorkspaceContentAuditDecision, DecisionNoOpinion,
+			WorkspaceContentAuditReason, "only root serviceaccounts may access root",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:135]: same as above, this is a possible resolution, the reason is rather something like `toplevel.authorization.kcp.io/reason: root workspace access by non-root service account not permitted`.

File: pkg/authorization/workspace_content_authorizer.go [hunk 12]
```diff
@@ -129,12 +168,28 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
 	ws, err := a.clusterWorkspaceLister.Get(clusters.ToClusterAwareKey(parentClusterName, cluster.Name.Base()))
 	if err != nil {
 		if errors.IsNotFound(err) {
+			kaudit.AddAuditAnnotations(
+				ctx,
+				WorkspaceContentAuditDecision, DecisionDenied,
+				WorkspaceContentAuditReason, "clusterworkspace not found",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:174]: fmt.Sprintf(`parent clusterworkspace %q not found`, parentClusterName)

File: pkg/authorization/workspace_content_authorizer.go [hunk 13]
```diff
@@ -178,22 +233,43 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
 			return authorizer.DecisionNoOpinion, strings.Join(reasonList, "\n"), utilerrors.NewAggregate(errList)
 		}
 	}
 
 	// non-admin subjects don't have access to initializing workspaces.
 	if ws.Status.Phase == tenancyv1alpha1.ClusterWorkspacePhaseInitializing && !extraGroups.Has(bootstrap.SystemKcpClusterWorkspaceAdminGroup) {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			WorkspaceContentAuditDecision, DecisionNoOpinion,
+			WorkspaceContentAuditReason, "only admin subjects may access initializing workspaces",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:250]: same as above, this is a possible resolution, the reason is rather something like `toplevel.authorization.kcp.io/reason: not permitted, clusterworkspace is in initializing phase`.

File: pkg/authorization/workspace_content_authorizer.go [hunk 14]
```diff
@@ -178,22 +233,43 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
+		)
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, nil
 	}
 
 	if len(extraGroups) == 0 {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			WorkspaceContentAuditDecision, DecisionNoOpinion,
+			WorkspaceContentAuditReason, "subject has no extraGroups",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:259]: `not permitted, subject has not been granted any groups`

File: pkg/authorization/workspace_content_authorizer.go [hunk 15]
```diff
@@ -178,22 +233,43 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
 		return authorizer.DecisionNoOpinion, WorkspaceAcccessNotPermittedReason, nil
 	}
 
 	withGroups := deepCopyAttributes(attr)
 	withGroups.User.(*user.DefaultInfo).Groups = append(attr.GetUser().GetGroups(), extraGroups.List()...)
 
+	kaudit.AddAuditAnnotations(
+		ctx,
+		WorkspaceContentAuditDecision, DecisionAllowed,
+		WorkspaceContentAuditReason, "subject has extraGroups",
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:270]: fmt.Sprintf(`allowed with additional groups %v`, extraGroups)

File: pkg/authorization/toplevel_org_authorizer.go (hunk 6 above)
Comment (ncdc) [reply pkg/authorization/toplevel_org_authorizer.go:128]: Isn't it already there with the `tenancy.kcp.dev/workspace` annotation?

File: pkg/authorization/workspace_content_authorizer.go (hunk 10 above)
Comment (ncdc) [reply pkg/authorization/workspace_content_authorizer.go:102]: tenancy.kcp.dev/workspace annotation has it?

File: pkg/authorization/workspace_content_authorizer.go (hunk 12 above)
Comment (ncdc) [reply pkg/authorization/workspace_content_authorizer.go:174]: this is checking for the workspace itself, not the parent

File: pkg/authorization/workspace_content_authorizer.go (hunk 10 above)
Comment (s-urbaniak) [reply pkg/authorization/workspace_content_authorizer.go:102]: ah, perfect then 👍 just making sure we're having it in the audit log.

File: pkg/authorization/toplevel_org_authorizer.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/authorization/toplevel_org_authorizer.go:128]: if so, then perfect 👍

File: pkg/authorization/workspace_content_authorizer.go (hunk 12 above)
Comment (s-urbaniak) [reply pkg/authorization/workspace_content_authorizer.go:174]: ah yes, the parent was the prefix only in get, my bad.

**Comment by ncdc** [comment]
Hit #1693
/retest

**Comment by ncdc** [comment]
Something appears to be wrong with Prow. It keeps picking up the previous failure and re-reporting it as a new failure (without actually running new tests).

/retest

**Pull request merged**